go/vt/vttablet/tabletmanager/restore.go

```diff
@@ -84,27 +84,212 @@ func init() {
+		Logger:               logger,
+		Concurrency:          restoreConcurrency,
+		HookExtraEnv:         tm.hookExtraEnv(),
+		DeleteBeforeRestore:  deleteBeforeRestore,
+		DbName:               topoproto.TabletDbName(tablet),
+		Keyspace:             keyspace,
+		Shard:                tablet.Shard,
+		StartTime:            startTime,
+		DryRun:               dryRun,
+		Stats:                backupstats.RestoreStats(),
+		MysqlShutdownTimeout: mysqlShutdownTimeout,
+		AllowedBackupEngines: allowedBackupEngines,
+		RestoreToPos:         restoreToPos,
+		RestoreToTimestamp:   restoreToTimestamp,
+	}
+	var backupManifest *mysqlctl.BackupManifest
+	for {
+		backupManifest, err = mysqlctl.Restore(ctx, params)
+		if backupManifest != nil {
+			statsRestorePosition.Set(replication.EncodePosition(backupManifest.Position))
+			statsRestoreBackupTime.Set(backupManifest.BackupTime)
+		}
+		logger.Infof("Restore: got a restore manifest: %v, err=%v, waitForBackupInterval=%v", backupManifest, err, waitForBackupInterval)
+		if waitForBackupInterval == 0 {
+			break
+		}
+		// Break if no errors.
+		if err == nil {
+			break
+		}
+		// Retry a specific set of errors. The rest we handle immediately.
+		if slices.Contains([]error{mysqlctl.ErrNoBackup, mysqlctl.ErrNoCompleteBackup}, err) {
+			break
+		}
+		log.Infof("No backup found. Waiting %v (from -wait-for-backup-interval flag) to check again.", waitForBackupInterval)
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		case <-time.After(waitForBackupInterval):
+		}
+	}
+
+	// Handle error.
+	if err == nil && backupManifest == nil {
+		err = vterrors.New(vtrpcpb.Code_INTERNAL, "no backup manifest")
+	}
+	if err != nil && err != mysqlctl.ErrNoBackup {
+		return replicationAction, replicationPos, overrideTabletType, vterrors.Wrap(err, "can't restore backup")
+	}
+
+	// Choose replication action.
+	switch {
+	case params.DryRun:
+		// Do nothing here.
+		params.Logger.Infof("Dry run. No changes made")
+		replicationAction = replicationActionNone
+	case params.IsIncrementalRecovery():
+		// The whole point of point-in-time recovery is that we want to restore
+		// up to a given position, and to NOT proceed from that position. We
+		// want to disable replication and NOT let the replica catch up with the
+		// primary.
+		replicationAction = replicationActionDisable
+	case err == mysqlctl.ErrNoBackup:
+		// Starting with empty database.
+		// We just need to initialize replication
+		replicationAction = replicationActionInitialize
+	case keyspaceType == topodatapb.KeyspaceType_NORMAL:
+		// Reconnect to primary only for "NORMAL" keyspaces.
+		replicationAction = replicationActionStart
+		replicationPos = backupManifest.Position
+	default:
+	}
+
+	// Override tablet type.
+	if params.IsIncrementalRecovery() && !params.DryRun {
+		params.Logger.Infof("Restore: will set tablet type to DRAINED as this is a point in time recovery")
+		drained := topodatapb.TabletType_DRAINED
+		overrideTabletType = &drained
+	}
+
```

Comment:
previously, in `restoreDataLocked`, we reverted directly to the previous tablet type. now that is the responsibility of `restoreCommonInnerLocked`.